milestones/data.py

```diff
@@ -0,0 +1,373 @@
+
+    if content_key:
+        queryset = queryset.filter(coursecontentmilestone__content_id=unicode(content_key))
+
+    if relationship:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = internal.Milestone.objects.filter(
+            coursecontentmilestone__milestone_relationship_type=mrt.id,
+            active=True,
+        )
+
+    course_content_milestones = []
+    if len(queryset):
+        for milestone in queryset:
+            course_content_milestones.append(serializers.serialize_milestone(milestone))
+
+    return course_content_milestones
+
+
+def fetch_milestone_courses(milestone, relationship=None):
+    """
+    Retrieves the set of courses currently linked to the specified milestone
+    Optionally pass in 'relationship' (ex. 'fulfills') to filter down the set
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    queryset = internal.CourseMilestone.objects.filter(
+        milestone=milestone_obj,
+        active=True
+    ).select_related('milestone')
+
+    # if milestones relationship type found then apply the filter
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(
+            milestone_relationship_type=mrt.id,
+        )
+
+    # Assemble the response container
+    milestone_courses = []
+    if len(queryset):
+        for milestone in queryset:
+            milestone_courses.append(serializers.serialize_milestone_with_course(milestone))
+
+    return milestone_courses
+
+
+def fetch_milestone_course_content(milestone, relationship=None):
+    """
+    Retrieves the set of course content modules currently linked to the specified milestone
+    Optionally pass in 'relationship' (ex. 'fulfills') to filter down the set
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    queryset = internal.CourseContentMilestone.objects.filter(
+        milestone=milestone_obj,
+        active=True
+    ).select_related('milestone')
+
+    # if milestones relationship type found then apply the filter
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(
+            milestone_relationship_type=mrt.id,
+        )
+
+    # Assemble the response container
+    milestone_course_content = []
+    for milestone in queryset:
+        milestone_course_content.append(
+            serializers.serialize_milestone_with_course_content(milestone)
+        )
+
+    return milestone_course_content
+
+
+def create_user_milestone(user, milestone):
+    """
+    Inserts a new user-milestone into app/local state
+    No response currently defined for this operation
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    internal.UserMilestone.objects.get_or_create(
+        user_id=user['id'],
+        milestone=milestone_obj,
+        active=True,
+    )
```

Comment:
So the question with "returning" in these particular workflows is what exactly to return.  For example, in create_user_milestone() we're providing the user and the milestone, and the system is creating a link between these two entities.  Should we then return the new UserMilestone relationship that was created?  Return the user?  The milestone?  Maybe we simply return True?

After we've decided on the proper pattern I can implement across the board -- I will say again that up until now I've had no reason to work with a response from these particular workflows, versus, say, the one that actually adds a new Milestone.